### PR TITLE
SQL injection detection

### PR DESCRIPTION
**Pull request by CaspianA1**
Mostly done, but not quite; I just made a PR to run e2e tests and get feedback on my progress so far.

Here's a summary of my changes:
1. Improved global variable mutation detection (can now detect weird variable swaps, like `z = [y, y = z][0];`, and modifications separated by a comment, like `x = 23 + x, y.a = 24 + x;`). This covers all assignment expression types, I believe, except for spread assignments, which I will need to implement at some point.
2. Minor function name changes, README changes, stuff like that
3. Lots of SQL injection detection code
4. Add lots of test cases for this
5. Some dependency tweaks, to make the chances of broken installs of this package yet smaller

Here's how I determine probable SQL injections:
- If I detect a callsite where raw SQL queries can happen (e.g. `ctxt.client.raw(myQuery)`, where the ORM client is `Knex`), figure out if the value is reducible to a string literal, or concatenations of string literals.
- This means that the parameter can be a string literal, a variable that is only ever assigned to string literals, a concatenation of string literals, a variable that's assigned to concatenations of string literals with other such variables, etc...
- Template expressions (e.g. ``` `${foo}, ${bar}, ${baz}` ```) are also supported, as long as the values used to build it (`foo`, `bar`, and `baz` here) are also reducible to string literals. This does not include tagged template expressions, since those involve a function call, and my variable definition analysis stays within the scope of a single `Transaction` function.
- Using any of a `Transaction` function's parameters for a raw SQL query is not allowed either, since you don't know who will call that transaction except in certain cases.
- If the passed-in variable to the query is not local, I examine the whole scope of the program for any non-literal-reducible assignments. If I find any, then I raise an error.
- If some variable passed into a callsite for raw SQL queries does not meet the aforementioned requirements, I flag it as a SQL injection, because that then means that you most likely took some datatype and cast it to a string to build your query, which is risky.

Here's what's left:
- [x] Fixing some versioning hassle
- [x] Add support for more callsites (beyond just `Knex`'s `raw` function). Should be relatively easy.
- [x] Proper line reporting for errors (provide the line where the potential injection happened, not the line of the raw SQL call). Medium difficulty (not particularly hard, it'd just take an hour or two to implement).
- [x] Dealing with type aliasing

And things to think about after that:
- [ ] Maybe split everything into more files (but that might be another PR)
- [x] Version bump

### CODE REVIEW

File: dbos-rules.ts [hunk 1]
```diff
@@ -262,32 +255,198 @@ const awaitsOnNotAllowedType: DetChecker = (node, _fn, _isLocal) => {
...
+    const subexpr = node.getExpression();
+
+    // `ctxt.client.<callName>`
+    const identifiers = subexpr.getDescendantsOfKind(SyntaxKind.Identifier);
 
-// At the moment, this only performs analysis on expected-to-be-deterministic functions
-function analyzeFunction(fn: FunctionOrMethod) {
-  const body = fn.getBody();
+    // An injection in DBOS must match `ctxt.client.<something>`
+    if (identifiers.length !== 3) return;
```
Comment (qianl15) [inline dbos-rules.ts]: Does this work if I assign `ctxt.client` to another variable and invoke `raw`?
```
const client = ctxt.client;
client.raw(`sql stmt`);
```

Comment (CaspianA1) [reply dbos-rules.ts]: Not yet - I have a TODO for that

File: dbos-rules.test.ts [hunk 2]
```diff
@@ -55,28 +55,244 @@ function makeExpectedDetCode(
...
+        const foo = "xyz", bar = "xyw";
+        ctxt.client.raw(foo);
+        ctxt.client.raw(bar);
+
+        // Variables -> variables -> ... -> literals
+
+        // Literal concatenation (TODO: make adding parentheses in not fail)
+        ctxt.client.raw("foo" + "bar" + "baz" + "bam");
+
+        // Literal + variable concatenation
```
Comment (chuck-dbos) [inline dbos-rules.test.ts:164]: I would point out that this is OK because those variables have known values.

**Comment by chuck-dbos** [comment]
Another thing on the future thought list... there are sanitization functions whose value can then be concatenated.  (Though usually easier just to use template, this comes up sometimes in DDLs.)

File: dbos-rules.test.ts (hunk 2 above)
Comment (CaspianA1) [reply dbos-rules.test.ts:164]: Yes, it is - also, I fixed the parentheses thing, haven't pushed that yet

File: dbos-rules.ts [hunk 3]
```diff
@@ -348,28 +678,36 @@ function analyzeRootNode(eslintNode: EslintNode, eslintContext: EslintContext) {
...
+  if (testingValidityOfTestsLocally) checkDiagnostics(tsMorphNode);
 
   try {
     if (Node.isStatemented(tsMorphNode)) {
       tsMorphNode.getFunctions().forEach(analyzeFunction);
       tsMorphNode.getClasses().forEach(analyzeClass);
     }
     else {
-      throw new Error(`Was expecting a statemented node! Got this kind instead: ${tsMorphNode.getKindName()}`);
+      const possibleVersioningError = `\
```
Comment (kraftp) [inline dbos-rules.ts]: As discussed, we actually care about the version of the typescript compiler used by `typescript-eslint`, which constructs the AST. We should figure out a range of compatible versions (those using TS 5.4.x, for now) and fail on startup if the app uses any version of `typescript-eslint` out of that range.

**Review by kraftp — APPROVED** [review]

**Pull request merged**